Add application credentials platform

allenporter opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Initial implementation of a OAuth application credentials platform, from [architecture discussion](https://github.com/home-assistant/architecture/discussions/692#discussioncomment-2121633), initially supporting:
- [X] Platform for integrations to provide authorization server information, and backwards compatibility for yaml based credentials
- [X] Websocket list/add/delete
- [X] Add developer credentials protocol from yaml config
- [X] Handle OAuth credential registration and de-registration
- [X] Tests for websocket and integration based registration
- [X] Support for XBOX as initial test case

Upcoming considerations:
- [ ] Updated scaffold script and developer documentation
- [x] "Import" and migration path for credentials
- [ ] Move existing local auth implementations to application credentials
- [ ] APIs for lower level manipulation for more complex cases (e.g. for nest)
- [ ] APIs needed for improved user experience (e.g. tighter config flow integration, smoother add/remove)
- [x] Frontend changes to show the user interface (https://github.com/home-assistant/frontend/pull/12344)

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/developers.home-assistant/pull/1279

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [X] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

allenporter committed:
Initial developer credentials scaffolding
- Support websocket list/add/delete
- Add developer credentials protocol from yaml config
- Handle OAuth credential registration and de-registration
- Tests for websocket and integration based registration

allenporter committed:
Fix pydoc text

balloob commented on CODEOWNERS:
```diff
@@ -219,6 +219,8 @@ build.json @home-assistant/supervisor
 /tests/components/denonavr/ @ol-iver @starkillerOG
 /homeassistant/components/derivative/ @afaucogney
 /tests/components/derivative/ @afaucogney
+/homeassistant/components/developer_credentials/ @allenporter
```
Let's mark it `@home-assistant/core`

allenporter replied:
👍🏼

balloob commented on homeassistant/components/developer_credentials/__init__.py:
```diff
@@ -0,0 +1,270 @@
+) -> None:
+    """Register an OAuth2 flow implementation for an integration."""
+    auth_domain = _get_auth_domain(domain, developer_credential)
+    unsub = hass.data[DOMAIN][domain][auth_domain]
```
Remove the key too:
```suggestion
    unsub = hass.data[DOMAIN][domain].pop(auth_domain)
```

balloob commented on homeassistant/components/developer_credentials/__init__.py:
```diff
@@ -0,0 +1,270 @@
+) -> None:
+    """Register an OAuth2 flow implementation for an integration."""
+    auth_domain = _get_auth_domain(domain, developer_credential)
+    authorization_server = await _async_get_authorization_server(hass, domain)
```
Let's raise if already registered. Shouldn't happen so good to catch.
```suggestion
    if auth_domain in hass.data[DOMAIN][domain]:
        raise ValueError(f"Domain {auth_domain} already registered")
    authorization_server = await _async_get_authorization_server(hass, domain)
```

balloob commented on homeassistant/components/developer_credentials/__init__.py:
```diff
@@ -0,0 +1,270 @@
+    hass: HomeAssistant, domain: str
+) -> DeveloperCredential | None:
+    """Return the DeveloperCredential for the integration domain."""
+    if domain not in hass.data[DOMAIN]:
```
Since this is a private method, this can't happen ? Or would this method be called by something that is called by other integrations ?

allenporter replied:
Yeah, ok, this method isn't doing anything useful. I'm inlining it where it was called.  (I think this had other uses originally that were removed as i developed it)

balloob commented on homeassistant/components/developer_credentials/translations/en.json:
```diff
@@ -0,0 +1,21 @@
+{
```
This file can be removed.

allenporter replied:
Removed.

allenporter committed:
Update homeassistant/components/developer_credentials/__init__.py

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

balloob commented on homeassistant/components/developer_credentials/__init__.py:
```diff
@@ -0,0 +1,270 @@
+    get_developer_credential = hass.data[DOMAIN][domain][
+        DeveloperCredentialsType.CONFIG_CREDENTIAL.value
+    ]
+    return await get_developer_credential(hass)
```
`get_developer_credential` can be None

allenporter replied:
I removed this.

allenporter committed:
Remove translations and update owners

balloob commented on homeassistant/components/developer_credentials/__init__.py:
```diff
@@ -0,0 +1,270 @@
+    # Register an authentication implementation for every credential provided
+    if get_config_developer_credential is None:
+        return
+    developer_credential = await _async_get_developer_credential(
```
There is a race condition here.

Integrations that want to use this integration need to add it to `dependencies`. That means that it's already set up by the time the integration that needs it is being set up.

The `async_process_integration_platforms` helper listens to `EVENT_COMPONENT_LOADED` to decide to check if the platform exists and register it. 

However the `EVENT_COMPONENT_LOAD` will not fire until the config entries are finished being set up ([source](https://github.com/home-assistant/core/blob/dev/homeassistant/setup.py#L268-L281)).

allenporter replied:
OK, so the order we want is:
- developer credentials setup
- integration setup
- developer credentials platform setup for that integration
- integration config entry

Is that right?

allenporter replied:
With my current limited understanding: this appears to be a fundamental mismatch for a platform, since it needs to run before config entries are created. I am not aware of other use cases that act line this.  Might it make sense to have integrations register in setup rather than using a platform? Or is this something worthwhile to incorporate into a platform feature. Want to get some feedback before acting.

balloob replied:
Maybe we can make the `config_entry_oauth2_flow.py` helper aware of the application_credentials integration. When ie Nest calls `config_entry_oauth2_flow.async_get_implementations(hass, "nest")` it will try to load the `application_credential` platform for Nest if `application_credential` integration is set up?

allenporter replied:
Ah looks like `config_entry_oauth2_flow.async_add_implementation_provider` may be able to assist with that. I'll have a closer look.

allenporter replied:
OK, i changed this to load via the implementation provider, with some updates to that API. This is nice because there is no logic for registering and de-registering, since everything is loaded as needed from storage.

I also added a first stab at import support since that came up in the docs PR.

allenporter committed:
Update homeassistant/components/developer_credentials/__init__.py

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

allenporter committed:
Remove _async_get_developer_credential

frenck commented on homeassistant/components/developer_credentials/manifest.json:
```diff
@@ -0,0 +1,9 @@
+{
+  "domain": "developer_credentials",
+  "name": "Developer Credentials",
```
I would opt to remove/avoid the term "developer". It might sound scary to less technical end-users. In OAuth specs, this is called "Client Credentials" and more often referred to as "Application Credentials" too (as most platforms, like Spotify, Twitter, and more, you create "Applications" on their portals).

Left the comment here as well: <https://github.com/home-assistant/architecture/discussions/692#discussioncomment-2494857>

allenporter replied:
I thought i had resolved this and just unresolved to make sure we circle back and are happy with the naming. (I realize now maybe i didn't resolve it, but i can't remember)

frenck replied:
Yeah, I think it is fine and fitting. Anything else than "developer" would be good 👍

allenporter committed:
Rename to application credentials platform

allenporter committed:
Fix race condition and add import support

balloob commented on homeassistant/components/application_credentials/__init__.py:
```diff
@@ -100,121 +97,90 @@ async def async_delete_item(self, item_id: str) -> None:
 
-    # Allow future registration of local oauth implementations
     config_entry_oauth2_flow.async_register_local_apis(hass)
+    config_entry_oauth2_flow.async_add_implementation_provider(
```
Yes, of course!

allenporter committed:
Increase code coverage (92%)

allenporter committed:
Increase test coverage 93%

allenporter committed:
Increase test coverage (94%)

allenporter committed:
Increase test coverage (97%)

allenporter committed:
Increase test covearge (98%)

allenporter committed:
Increase test coverage (99%)

allenporter committed:
Increase test coverage (100%)

balloob commented on homeassistant/helpers/config_entry_oauth2_flow.py:
```diff
@@ -329,17 +329,29 @@ def async_register_implementation(
+def async_register_local_apis(hass: HomeAssistant) -> None:
+    """Initialize requirements for local OAuth2 implementations.
+
+    This must be called on startup before the http router is frozen.
```
We prevent the router from getting frozen so this is not an issue to work around.

allenporter replied:
I am having trouble reconciling your comment with the error message I see:
```
homeassistant/data_entry_flow.py:205: in async_init
    flow, result = await task
homeassistant/data_entry_flow.py:232: in _async_init
    result = await self._async_handle_step(flow, flow.init_step, data, init_done)
homeassistant/data_entry_flow.py:335: in _async_handle_step
    result: FlowResult = await getattr(flow, method)(user_input)
homeassistant/helpers/config_entry_oauth2_flow.py:319: in async_step_user
    return await self.async_step_pick_implementation(user_input)
homeassistant/helpers/config_entry_oauth2_flow.py:236: in async_step_pick_implementation
    implementations = await async_get_implementations(self.hass, self.DOMAIN)
homeassistant/helpers/config_entry_oauth2_flow.py:370: in async_get_implementations
    async_register_local_apis(hass)
homeassistant/helpers/config_entry_oauth2_flow.py:349: in async_register_local_apis
    hass.http.register_view(OAuth2AuthorizeCallbackView())
homeassistant/components/http/__init__.py:295: in register_view
    view.register(self.app, self.app.router)
homeassistant/components/http/view.py:95: in register
    routes.append(router.add_route(method, url, handler))
venv/lib/python3.9/site-packages/aiohttp/web_urldispatcher.py:1100: in add_route
    resource = self.add_resource(path, name=name)
venv/lib/python3.9/site-packages/aiohttp/web_urldispatcher.py:1085: in add_resource
    self.register_resource(resource)

...
resource = <PlainResource  /auth/external/callback>

    def register_resource(self, resource: AbstractResource) -> None:
        assert isinstance(
            resource, AbstractResource
        ), f"Instance of AbstractResource class is required, got {resource!r}"
        if self.frozen:
>           raise RuntimeError("Cannot register a resource into frozen router.")
E           RuntimeError: Cannot register a resource into frozen router.
```

My impression is that we are in fact freezing the router and this registration must be done at startup time, rather than just in time when the first local oauth implementation is registered.  Should I interpret your comment to mean the test harness is incorrect?

allenporter replied:
The above error happens if i do the API registration lazily, like how the non-provider does it for local oauth implementations.  An alternative framing is that the integration just needs to call this to have the APIs exposed at all when using the provider interface, so i removed the comment.

balloob replied:
https://github.com/home-assistant/core/blob/dev/homeassistant/components/http/__init__.py#L441-L446

allenporter replied:
Well .. that nice "creative" :) solution to avoid freezing does not run during tests. I am not going to dig into this, as moving the registration to the a component seems like the best fix. I'll revert changes to this file in this PR.

balloob replied:
Yeah this is one of those ticking time bombs that will one aiohttp update explode in our face. I think that we can workaround it using a resource, [like we did for the frontend index](https://github.com/home-assistant/core/blob/dev/homeassistant/components/frontend/__init__.py#L503), but never dug in to solve it.

balloob commented on homeassistant/components/application_credentials/__init__.py:
```diff
@@ -0,0 +1,215 @@
+) -> None:
+    """Import an existing credential from configuration.yaml.
+
+    The imported credential auth domain should match the existing auth implementation domain used
```
Isn't this always `domain` for now ?

allenporter replied:
It won't be for nest, but also nest may not be able to use this import flow anyway. I'll remove and revisit when we get to making this work for nest.

balloob commented on homeassistant/components/application_credentials/__init__.py:
```diff
@@ -0,0 +1,215 @@
+        CONF_CLIENT_SECRET: credential.client_secret,
+        CONF_AUTH_IMPL: auth_domain,
+    }
+    await storage_collection.async_create_item(item)
```
Shouldn't we deduplicate to avoid importing a credential multiple times?

allenporter replied:
Done. I made a call to ignore existing entries, though this means that a changed client secret would NOT be imported (e.g you'd need to go remove from the UI if you want to change the value). This seemed nice and simple, but i'm still calibrating my judgement w.r.t. trade-offs for end users of the project so can also make this overwrite if that is what we'd prefer.

balloob commented on homeassistant/components/application_credentials/__init__.py:
```diff
@@ -0,0 +1,215 @@
+            err,
+        )
+        return None
+    _LOGGER.debug(hasattr(platform, "async_get_authorization_server"))
```
stale debug.

balloob commented on homeassistant/components/application_credentials/manifest.json:
```diff
@@ -0,0 +1,9 @@
+  "name": "Application Credentials",
+  "config_flow": false,
+  "documentation": "https://www.home-assistant.io/integrations/application_credentials",
+  "dependencies": ["http"],
```
Why does it depend on `http` ?

allenporter replied:
It needs the OAuth HTTP callback APIs. My impression was that integrations should be responsible for the HTTP APIs provided the OAuth helpers since OAuth config flow is not a component, rather than hoping for the best.

Is it a better practice to assume that some other component will register this and just make it a test only dependency?

```
ERROR    homeassistant.setup:setup.py:245 Error during setup of component application_credentials
Traceback (most recent call last):
  File "/home/allen/home-assistant-core/homeassistant/setup.py", line 235, in _async_setup_component
    result = await task
  File "/home/allen/home-assistant-core/homeassistant/components/application_credentials/__init__.py", line 127, in async_setup
    config_entry_oauth2_flow.async_register_local_apis(hass)
  File "/home/allen/home-assistant-core/homeassistant/helpers/config_entry_oauth2_flow.py", line 349, in async_register_local_apis
    hass.http.register_view(OAuth2AuthorizeCallbackView())
AttributeError: 'NoneType' object has no attribute 'register_view'
```

balloob replied:
Good point. Yeah it should depend on it.

Or actually, we should register `OAuth2AuthorizeCallbackView` maybe in the `auth` component? It's weird that it's dynamically loaded.

allenporter replied:
I'll "pre-factor" this into a separate PR right now.

allenporter replied:
Sent https://github.com/home-assistant/core/pull/69507

allenporter replied:
I rebased and removed the changes around http registration, and updated this to depend on auth. All the complexity around registration is gone now.

allenporter committed:
Remove http router frozen comment

allenporter committed:
Remove auth domain override on import

allenporter committed:
Remove debug statement

allenporter committed:
Don't import the same client id multiple times

allenporter committed:
Add auth dependency for local oauth implementation

allenporter committed:
Revert older oauth2 changes from merge

MartinHjelmare commented on homeassistant/components/application_credentials/__init__.py:
```diff
@@ -0,0 +1,213 @@
+it.
+
+Integrations register an authorization server, and then the APIs are used to add
+add one or more client credentials. Integrations may also provide credentials
```
```suggestion
one or more client credentials. Integrations may also provide credentials
```

MartinHjelmare commented on homeassistant/components/application_credentials/__init__.py:
```diff
@@ -0,0 +1,213 @@
+
+This integration provides APIs for managing local OAuth credentials on behalf of other
+integrations. The preferred approach for all OAuth integrations is to use the cloud
+account linking service, and this is the alternative for integrations that can't use
```
This isn't clear, since we don't implement cloud account linking on its own. We support a local app by default and cloud account linking is added after that.

allenporter replied:
Removed.

MartinHjelmare commented on homeassistant/components/application_credentials/__init__.py:
```diff
@@ -0,0 +1,213 @@
+
+    CREATE_SCHEMA = vol.Schema(CREATE_FIELDS)
+
+    async def _process_create_data(self, data: dict) -> dict:
```
Do we know what type the keys are in the data dict?

allenporter replied:
This is effectively `CREATE_FIELDS` schema -- so dict[str, str] for now. I've updated to use that.

MartinHjelmare commented on homeassistant/components/application_credentials/__init__.py:
```diff
@@ -0,0 +1,213 @@
+        result = self.CREATE_SCHEMA(data)
+        domain = result[CONF_DOMAIN]
+        if not await _get_platform(self.hass, domain):
+            raise ValueError("No application_credentials platform for %s" % domain)
```
Please use f-strings.

allenporter replied:
Updated.

MartinHjelmare commented on homeassistant/components/application_credentials/__init__.py:
```diff
@@ -0,0 +1,213 @@
+        return result
+
+    @callback
+    def _get_suggested_id(self, info: dict) -> str:
```
Do we know the key type of the info dict?

allenporter replied:
Updated typing.

MartinHjelmare commented on tests/components/application_credentials/test_init.py:
```diff
@@ -0,0 +1,598 @@
+            async_get_authorization_server=AsyncMock(return_value=authorization_server),
+        ),
+    )
+    if config_credential:
```
I'd move this case to an explicit fixture or test.

MartinHjelmare commented on tests/components/application_credentials/test_init.py:
```diff
@@ -0,0 +1,598 @@
+    config_credential: ClientCredential | None,
+):
+    """Mock a application_credentials integration."""
+    if disable_setup:
```
Override the mock_application_credentials_integration fixture instead per test, via pytest.mark.parametrize.

allenporter replied:
Fixed.

allenporter committed:
Update homeassistant/components/application_credentials/__init__.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

allenporter committed:
Move config credential import to its own fixture

allenporter committed:
Override the mock_application_credentials_integration fixture instead per test

allenporter committed:
Update application credentials

allenporter committed:
Add dictionary typing

allenporter committed:
Use f-strings as per feedback

allenporter commented:
> Are there any frontend changes needed to support this?

There will need to be, yes. I think we should not start importing any credentials, for example, until users have the ability to go delete them in the UI if they are orphaned.  I have not thought through them in much detail yet, but  there will likely need additional APIs e.g. list of integrations supporting credentials.

MartinHjelmare commented:
How is an application added now in the frontend?

allenporter commented:
It's not yet, but it is needed before this can be actually used by users. My intention was to add a few pieces at a time in a way that was safe. (E.g. nobody will call these APIs yet)

I originally considered the UI orthogonal from the lower level APIs.  My intention originally was not to support import to storage at first for example, as we could move to the new library/platform with no new UI. (Now that import writes to storage we'll need a UI, or I go back to yaml creds that are configured, but not persisted)

Apologies if this wasn't clear but this is what I meant by "initial" in the PR title with a list of upcoming additions in the PR description.

MartinHjelmare commented:
Ok, we normally don't merge the backend PR, if it requires frontend changes to reach MVP, until we have a frontend PR and we know that it has what it needs from the backend.

allenporter commented:
Ok, sure, it definitely needs more.

allenporter committed:
Add additional structure needed for an MVP application credential

Add additional structure needed for an MVP, including a target
component Xbox

allenporter committed:
Add websocket to list supported integrations for frontend selector

allenporter committed:
Application credentials config

allenporter commented:
I need to now beef up tests for xbox as well and confirm that import works.

allenporter committed:
Import xbox credentials

allenporter committed:
Remove unnecessary async calls

allenporter commented on homeassistant/components/application_credentials/__init__.py:
```diff
@@ -218,3 +223,14 @@ async def _get_platform(
+    {vol.Required("type"): "application_credentials/integrations/list"}
+)
+@websocket_api.async_response
+async def handle_integration_list(
```
FIxed, now is a callback non-async function.

allenporter commented:
> I need to now beef up tests for xbox as well and confirm that import works.

Added creds import for xbox, which passes tests w/o any modification.  I picked xbox instead of google since it uses a plain LocalOAuth implementation (unlike nest and google calendar)

allenporter commented:
I've confirmed we have everything we need in the Frontend PR 
 https://github.com/home-assistant/frontend/pull/12344 (draft, waiting for this to be merged, but had a couple rounds of initial feedback)

allenporter commented:
Not pushing hard to get this in beta since I think we want broader integration support.

MartinHjelmare commented on script/hassfest/application_credentials.py:
```diff
@@ -0,0 +1,64 @@
+""".strip()
+
+
+def generate_and_validate(integrations: dict[str, Integration], config: Config):
```
```suggestion
def generate_and_validate(integrations: dict[str, Integration], config: Config) -> str:
```

MartinHjelmare commented on script/hassfest/application_credentials.py:
```diff
@@ -0,0 +1,64 @@
+    return BASE.format(json.dumps(match_list, indent=4))
+
+
+def validate(integrations: dict[str, Integration], config: Config):
```
```suggestion
def validate(integrations: dict[str, Integration], config: Config) -> None:
```

MartinHjelmare commented on script/hassfest/application_credentials.py:
```diff
@@ -0,0 +1,64 @@
+    if config.specific_integrations:
+        return
+
+    with open(str(application_credentials_path)) as fp:
+        if fp.read().strip() != content:
+            config.add_error(
+                "application_credentials",
+                "File application_credentials.py is not up to date. Run python3 -m script.hassfest",
+                fixable=True,
+            )
+        return
```
```suggestion
    if application_credentials_path.read_text(encoding="utf-8").strip() != content:
        config.add_error(
            "application_credentials",
            "File application_credentials.py is not up to date. Run python3 -m script.hassfest",
            fixable=True,
        )
```

MartinHjelmare commented on script/hassfest/application_credentials.py:
```diff
@@ -0,0 +1,64 @@
+    application_credentials_path = (
+        config.root / "homeassistant/generated/application_credentials.py"
+    )
+    with open(str(application_credentials_path), "w") as fp:
+        fp.write(f"{config.cache['application_credentials']}\n")
```
```suggestion
    application_credentials_path.write_text(
        f"{config.cache['application_credentials']}\n", encoding="utf-8"
    )
```

allenporter committed:
Update script/hassfest/application_credentials.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

allenporter committed:
Update script/hassfest/application_credentials.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

allenporter committed:
Update script/hassfest/application_credentials.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

allenporter committed:
Update script/hassfest/application_credentials.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

MartinHjelmare reviewed (approved):
Test failure looks unrelated. Maybe rebase?

MartinHjelmare reviewed (approved):
Great! Can be merged when frontend PR is happy with this.

allenporter commented:
> Great! Can be merged when frontend PR is happy with this.

Thanks for all the review time on this.

https://github.com/home-assistant/frontend/pull/12344#issuecomment-1110345495
"When the backend gets merged can you set it back to ready for review?"

:) I'll poke and see if frontend is close.

allenporter committed:
Import credentials with a fixed auth domain

Resolve an issue with compatibility of exisiting config entries when importing
client credentials

allenporter commented:
Frontend PR is good to go.

I updated how importing works to register imported credentials with the specified auth domain, rather than using the unique id auth domain based on the client id. This is needed for compatibility with existing config entries and was found when i was testing with other integrations besides xbox (future PRs).  Have another look.
